engine/src/main/java/org/terasology/rendering/nui/widgets/TypeWidgetFactoryRegistryImpl.java

```diff
@@ -0,0 +1,81 @@
+import com.google.common.collect.Maps;
+import org.terasology.assets.management.AssetManager;
+import org.terasology.config.flexible.ui.AutoConfigWidgetFactory;
+import org.terasology.context.Context;
+import org.terasology.engine.module.ModuleManager;
+import org.terasology.nui.widgets.types.TypeWidgetFactory;
+import org.terasology.nui.widgets.types.TypeWidgetFactoryRegistry;
+import org.terasology.nui.widgets.types.TypeWidgetLibrary;
+import org.terasology.nui.widgets.types.builtin.BooleanWidgetFactory;
+import org.terasology.nui.widgets.types.builtin.ByteWidgetFactory;
+import org.terasology.nui.widgets.types.builtin.DoubleWidgetFactory;
+import org.terasology.nui.widgets.types.builtin.FloatWidgetFactory;
+import org.terasology.nui.widgets.types.builtin.IntegerWidgetFactory;
+import org.terasology.nui.widgets.types.builtin.LongWidgetFactory;
+import org.terasology.nui.widgets.types.builtin.ShortWidgetFactory;
+import org.terasology.nui.widgets.types.builtin.StringWidgetFactory;
+import org.terasology.persistence.typeHandling.InstanceCreator;
+import org.terasology.reflection.reflect.ConstructorLibrary;
+import org.terasology.registry.InjectionHelper;
+import org.terasology.rendering.nui.widgets.types.ArrayWidgetFactory;
+import org.terasology.rendering.nui.widgets.types.CollectionWidgetFactory;
+import org.terasology.rendering.nui.widgets.types.EnumWidgetFactory;
+import org.terasology.rendering.nui.widgets.types.object.ObjectWidgetFactory;
+
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Registers {@link TypeWidgetFactory} instances that can be used by a {@link TypeWidgetLibrary}
+ * to generate widgets to edit objects of various types.
+ */
+public class TypeWidgetFactoryRegistryImpl implements TypeWidgetFactoryRegistry {
+    private final Context context;
+    private final List<TypeWidgetFactory> factories = new ArrayList<>();
+
+    private final Map<Type, InstanceCreator<?>> instanceCreators = Maps.newHashMap();
+    private final ConstructorLibrary constructorLibrary = new ConstructorLibrary(instanceCreators);
+
+    public TypeWidgetFactoryRegistryImpl(Context context) {
+        this.context = context;
+        addFactoriesForBuiltinTypes();
+    }
+
+    private void addFactoriesForBuiltinTypes() {
+        add(new ObjectWidgetFactory());
+
+        add(new BooleanWidgetFactory());
+
+        add(new ByteWidgetFactory());
+        add(new ShortWidgetFactory());
+        add(new IntegerWidgetFactory());
+        add(new LongWidgetFactory());
+        add(new FloatWidgetFactory());
+        add(new DoubleWidgetFactory());
+
+        add(new StringWidgetFactory());
+
+        add(new EnumWidgetFactory());
+        add(new CollectionWidgetFactory(constructorLibrary));
+        add(new ArrayWidgetFactory(constructorLibrary));
+        add(new AutoConfigWidgetFactory(context.get(ModuleManager.class), context.get(AssetManager.class)));
```

Comment:
Let's not do this here; let's instead use the annotation and inject the `ModuleManager` and `AssetManager` dependencies via `@In` _or_ create and add the `AutoConfigWidgetFactory` to the registry in the initialization phase of `AutoConfigManager`. This will make life easier when we move the type widget library into TeraNUI and extract `AutoConfig` into its own separate repository.

Reply:
I too :D

But our current DI too weak. 
Not supports: Collection Inject (best way for Libraries and generators), jsr310.

BestPractics for DI-world  - use constructor or setters don't break incapsulation.

aaand we have gestalt-di PR which provides nice DI implementation inspired by Spring and Micronauts.
By my opinion it should instroduce in DS firsts.
(also i have some PRs for DI, which prefers Constructor inject and using jsr310(`@Inject`)and spread DI from heard of engine)